src/core/dev_api/tensor_data_accessor.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "tensor_data_adapter.hpp"
+
+namespace ov {
+
+/** @brief Interface for data accessor. */
+class OPENVINO_API ITensorAccessor {
+public:
+    /**
+     * @brief Get tensor adapter at index
+     *
+     * @param port  Number of data port to get tensor adapter (operator input).
+     * @return      Constant pointer to tensor adapter interface.
+     */
+    virtual auto operator()(size_t port) const -> const ITensorDataAdapter* = 0;
+
+    virtual ~ITensorAccessor() = default;
+};
+
+/**
+ * @brief Null tensor accessor.
+ *
+ * Return null pointer for any input port.
+ */
+struct OPENVINO_API NullTensorAccessor : public ITensorAccessor {
+    const ITensorDataAdapter* operator()(size_t port) const override {
+        return nullptr;
+    }
+};
+
+/**
+ * @brief Get null data adapter pointer for any index number.
+ *
+ * @param port  Port number to get data.
+ * @return      Null pointer.
+ */
+inline auto null_tensor_accessor() -> const ITensorAccessor& {
+    static const auto null_accessor = NullTensorAccessor();
+    return null_accessor;
+};
+
+/**
+ * @brief Tensor data accessor functor.
+ *
+ * Creates the tensor data adapter if found in tensor container.
+ * This accessor not take ownership of tensors container.
+ * Supports following containers:
+ * - ov::TensorVector
+ * - ov::HostTensorVector
+ * - std::map<size_t, ov::HostTensorPtr>
+ *
+ * @tparam TContainer Type of tensor container.
+ */
+template <class TContainer>
+class OPENVINO_API TensorAccessor : public ITensorAccessor {
```

Comment:
The `ITensorAccessor` is to provide common way to access tensor-like data from containers (looks like map, and vector is commonly used). The shape_infer function then can access tensor data by requesting it by port index from plugin or core.
Main purpose for this accessor is bound check for data container and return `ITensorDataAdapter` to provide common way to access data as the `ITensor` is not done yet.

I was trying using the `ov::Tensor` instead but got some issues and also try follow @maxnick suggestion to do this access as simple as possible:

1. To avoid additional allocation. As plugin already has data so creating the ov::Tensor which holds at least one smart pointer and `ctor` are not very simple will cause additional allocations. The `DataAccessor` holds only the pointer to data adapter which should be as simple as possible (except GPU version it got only raw pointer). During accessing data the pointer value is replaced and returned. It should be cheap.
   - I think CPU plugin can do it similar as done for Tensors and creating new object will be not required.
   - I know that current version of data accessor for GPU use unique pointer (dynamic allocation) but I think can be optimized to use same pattern as for tensor to have only raw pointer copy but probably will required different approach to create memory lock. I think GPU team can optimize it for their needs.

2. In case of GPU plugin there is also required to create lock for accessed data. I don't know how e.g. create `ov::Tensor` and store the lock inside so this tensor adapter allow store lock and release when object will be deleted.

3. In case the ov::Tensor, I agree that adapter is overhead but the `ITensor` and `TensorView` classes from #14156 are not available yet. But I need common interface to access data. When `ITensor` will be don I think I can refactor and replace `ITensorDataAdapter` by `ITensor`. Then Accessor for `ov::Tensor` will be very simple as just bound check and return pointer or ref to data from tensor container.

4. In case of HostTensor it can be beneficial as access to data and conversion to adapter is raw pointer replace. Currently in CPU plugin the memories are converted to HostTensorPtr so it allow avoid again conversion to ov::Tensor just use pointers. If `HostTensor` implement `ITensor` then simplification will be same as for `ov::Tensor`.
If will not implement it then the `TensorView` could be used as adapter.

5. If plugin memories will implement `ITensor` then accessing data will be very simple if not maybe `TensorView` can be used to provide common interface.

I've put this files in dev API to share it between core, shape_inference, plugins and if possible share code from on lib. I can move it to shape_infer but then without exportin each library will have its own code.

I'm open to suggestions if there is possibility to have Tensor API implementation and avoid allocation etc. please share the idea.